common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/AbstractRootConcept.scala

```diff
@@ -15,6 +22,83 @@ object Concept {
   }
 }
 
+case class InstantRange(label: String,
+                        from: Instant,
+                        to: Instant,
+                        inferred: Boolean)
+    extends AbstractConcept
+object InstantRange {
+  // We use this apply as it's easier to work with date math on LocalDateTime than it is on Instant
+  def apply(label: String,
+            from: LocalDateTime,
+            to: LocalDateTime,
+            inferred: Boolean): InstantRange =
+    InstantRange(
+      label = label,
+      from = from.toInstant(ZoneOffset.UTC),
+      to = to.toInstant(ZoneOffset.UTC),
+      inferred = inferred)
+
+  type GetInstantRange = (String, LocalDateTime) => InstantRange
+  type InstantRangeParser = (DateTimeFormatter, GetInstantRange)
+  type DatePattern = String
+
+  val parsers: List[(DatePattern, GetInstantRange)] = List(
+    (
+      "yyyy",
+      (label: String, from: LocalDateTime) =>
+        InstantRange(
+          label,
+          from,
+          to = from.plusYears(1).minusNanos(1),
+          inferred = false)
+    ),
+    (
+      "'['yyyy']'",
+      (label: String, from: LocalDateTime) =>
+        InstantRange(
+          label,
+          from = from,
+          to = from.plusYears(1).minusNanos(1),
+          inferred = true)
+    )
+  )
+
+  // This explicitly defaults missing pieces to incomplete dates such as "1909" to
+  // then return 1909-01-01 to allow us to format it to the complete ISO8601 standard
+  private def formatterWithDefaults(pattern: String): DateTimeFormatter =
+    new DateTimeFormatterBuilder()
+      .appendPattern(pattern)
+      .parseDefaulting(ChronoField.NANO_OF_SECOND, 0)
+      .parseDefaulting(ChronoField.SECOND_OF_MINUTE, 0)
+      .parseDefaulting(ChronoField.MINUTE_OF_HOUR, 0)
+      .parseDefaulting(ChronoField.HOUR_OF_DAY, 0)
+      .parseDefaulting(ChronoField.DAY_OF_MONTH, 1)
+      .parseDefaulting(ChronoField.MONTH_OF_YEAR, 1)
+      .parseDefaulting(ChronoField.YEAR_OF_ERA, Year.now().getValue())
+      .toFormatter()
+
+  @tailrec
+  private def findParser(
+    label: String,
+    parsers: List[(DatePattern, GetInstantRange)]): Option[InstantRange] = {
+
+    parsers match {
+      case (pattern: DatePattern, getInstantRange: GetInstantRange) :: tail =>
+        val tryLocalDateTime = Try(
+          LocalDateTime.parse(label, formatterWithDefaults(pattern)))
+
+        if (tryLocalDateTime.isSuccess)
+          tryLocalDateTime.toOption.map(ldt => getInstantRange(label, ldt))
+        else findParser(label, tail)
```

Comment:
This sort of `if (tryValue.isSuccess) … else ` isn't very idiomatic Scala, or at least compared to the rest of our codebase! We do a lot of pattern matching, e.g.:

```scala
tryLocalDateTime match {
  case Success(localDateTime) => Some(getInstanceRange(label, localDateTime))
  case Failure(_) => findParser(label, tail)
}
```

Reply:
I knew I was missing something.